ffi/newpassmanagers.cpp

```diff
@@ -0,0 +1,319 @@
+    return llvm::unwrap(PTO)->SLPVectorization;
+}
+
+API_EXPORT(void)
+LLVMPY_PTOSetSLPVectorization(LLVMPipelineTuningOptionsRef PTO, bool value) {
+    llvm::unwrap(PTO)->SLPVectorization = value;
+}
+
+API_EXPORT(bool)
+LLVMPY_PTOGetLoopUnrolling(LLVMPipelineTuningOptionsRef PTO) {
+    return llvm::unwrap(PTO)->LoopUnrolling;
+}
+
+API_EXPORT(void)
+LLVMPY_PTOSetLoopUnrolling(LLVMPipelineTuningOptionsRef PTO, bool value) {
+    llvm::unwrap(PTO)->LoopUnrolling = value;
+}
+
+// FIXME: Available from llvm16
+// API_EXPORT(int)
+// LLVMPY_PTOGetInlinerThreshold(LLVMPipelineTuningOptionsRef PTO) {
+//     return llvm::unwrap(PTO)->InlinerThreshold;
+// }
+
+// API_EXPORT(void)
+// LLVMPY_PTOSetInlinerThreshold(LLVMPipelineTuningOptionsRef PTO, bool value) {
+//     llvm::unwrap(PTO)->InlinerThreshold = value;
+// }
+
+API_EXPORT(void)
+LLVMPY_DisposePipelineTuningOptions(LLVMPipelineTuningOptionsRef PTO) {
+    delete llvm::unwrap(PTO);
+}
+
+// PB
+
+API_EXPORT(LLVMPassBuilderRef)
+LLVMPY_CreatePassBuilder(LLVMTargetMachineRef TM,
+                         LLVMPipelineTuningOptionsRef PTO) {
+    TargetMachine *target = llvm::unwrap(TM);
+    PipelineTuningOptions *pt = llvm::unwrap(PTO);
+    return llvm::wrap(new PassBuilder(target, *pt));
+}
+
+API_EXPORT(void)
+LLVMPY_DisposePassBuilder(LLVMPassBuilderRef PB) { delete llvm::unwrap(PB); }
+
+static OptimizationLevel mapLevel(int speed_level, int size_level) {
+    switch (size_level) {
+    case 0:
+        switch (speed_level) {
+        case 0:
+            return OptimizationLevel::O0;
+        case 1:
+            return OptimizationLevel::O1;
+        case 2:
+            return OptimizationLevel::O2;
+        case 3:
+            return OptimizationLevel::O3;
+        default:
+            llvm_unreachable("Invalid optimization level");
+        }
+    case 1:
+        if (speed_level == 1)
+            return OptimizationLevel::Os;
+        llvm_unreachable("Invalid optimization level for size level 1");
+    case 2:
+        if (speed_level == 2)
+            return OptimizationLevel::Oz;
+        llvm_unreachable("Invalid optimization level for size level 2");
+    default:
+        llvm_unreachable("Invalid size level");
+        break;
+    }
+}
+
+API_EXPORT(LLVMModulePassManagerRef)
+LLVMPY_buildPerModuleDefaultPipeline(LLVMPassBuilderRef PBref, int speed_level,
+                                     int size_level) {
+
+    PassBuilder *PB = llvm::unwrap(PBref);
+    OptimizationLevel OL = mapLevel(speed_level, size_level);
+
+    // FXIME: No need to explicitly take care of O0 from LLVM 17
+    if (OL == OptimizationLevel::O0) {
+        return llvm::wrap(
+            new ModulePassManager(PB->buildO0DefaultPipeline(OL)));
+    }
+
+    return llvm::wrap(
+        new ModulePassManager(PB->buildPerModuleDefaultPipeline(OL)));
+}
+
+API_EXPORT(LLVMFunctionPassManagerRef)
+LLVMPY_buildFunctionSimplificationPipeline(LLVMPassBuilderRef PBref,
```

Comment:
I hope it's ok if I chime in a little here...
I think it might be helpful if the ModuleSimplificationPipeline was also exposed? (https://llvm.org/doxygen/classllvm_1_1PassBuilder.html#ad6f258d31ffa2d2e4dfaf990ba596d0d). That could be used to simplify whole modules before they are linked into multiple different modules?

I also wonder how hard it would be to expose some of the extension points for the pipelines (https://llvm.org/docs/NewPassManager.html#inserting-passes-into-default-pipelines) so for instance the `registerPipelineEarlySimplificationEPCallback` method on `PassBuilder`? Maybe that would be a nice way of adding the refpune pass to reasonable locations inside a pipeline?

Reply:
> I think it might be helpful if the ModuleSimplificationPipeline was also exposed?

BuildPerModuleDefultPipeline does call this internally, do you explicitly need it?

> I also wonder how hard it would be to expose some of the extension points for the pipelines

Won't be too difficult but it would make the APIs confusing I believe(maybe not, haven't really thought about in it in depth)

If you don't need them urgently, can you create separate tickets for them and I can revisit them again?

Reply:
`buildPerModuleDefultPipeline` calls `buildModuleSimplificationPipeline` ([here](https://github.com/llvm/llvm-project/blob/bc946f52870a51b49a0c7c3c508cef1905a11bc0/llvm/lib/Passes/PassBuilderPipelines.cpp#L1579)) but it *also* calls `buildModuleOptimizationPipeline`. And according to the docs that should not be executed repeatedly. So `per_module_default_pipeline(per_module_default_pipeline(module).link_into(module2))` would be incorrect, but `per_module_default_pipeline(module_simplification_pipeline(module1).link_into(module2))` would be fine, right? (That's one of the things I've been experimenting in https://github.com/numba/numba/pull/9566.)

About the register_* functions:
I'm only curious if they can be used to speed up numba compilation, because then fewer total optimization/simplification passes might be needed. Nothing urgent about it, it just might (?) make numba compilation faster (assuming the numba devs agree that that's a good idea anyway...).

Reply:
Created #1055 to further the discussion(s)